midp/media.js

```diff
@@ -215,70 +242,69 @@ AudioPlayer.prototype.realize = function() {
     return Promise.resolve(1);
 };
 
-AudioPlayer.prototype.play = function() {
-    this.audio.play();
-    this.audio.onended = function() {
-        MIDP.sendEndOfMediaEvent(this.playerContainer.pId, this.getDuration());
-    }.bind(this);
-};
-
 AudioPlayer.prototype.start = function() {
     if (this.playerContainer.contentSize == 0) {
         console.warn("Cannot start playing.");
         return;
     }
 
-    if (this.audio.src) {
-        this.play();
-        return;
+    var array = null;
+    if (!this.loaded) {
+        var data = this.playerContainer.data.subarray(0, this.playerContainer.contentSize);
+        // Convert the data to a regular Array to traverse the mozbrowser boundary.
+        var array = Array.prototype.slice.call(data);
+        array.constructor = Array;
+        this.loaded = true;
     }
-
-    new Promise(function(resolve, reject) {
-        var blob = new Blob([ this.playerContainer.data.subarray(0, this.playerContainer.contentSize) ],
-                            { type: this.playerContainer.contentType });
-        this.audio.src = URL.createObjectURL(blob);
-        this.audio.onloadedmetadata = function() {
-            resolve();
-            this.play();
-        }.bind(this);
-        this.audio.onerror = reject;
-    }.bind(this)).done(function() {
-        URL.revokeObjectURL(this.audio.src);
-    }.bind(this));
+    this.sender({
+        type: "start",
+        contentType: this.playerContainer.contentType,
+        data: array
+    });
+    this.paused = false;
 };
 
 AudioPlayer.prototype.pause = function() {
-    if (this.audio.paused) {
+    if (this.paused) {
         return;
     }
-    this.audio.onended = null;
-    this.audio.pause();
+    this.sender({ type: "pause" });
+    this.paused = true;
 };
 
 AudioPlayer.prototype.resume = function() {
-    if (!this.audio.paused) {
+    if (!this.paused) {
         return;
     }
-    this.play();
+    this.sender({ type: "play" });
+    this.paused = false;
 };
 
 AudioPlayer.prototype.close = function() {
-    this.audio.pause();
-    this.audio.src = "";
+    this.sender({ type: "close" });
+    this.paused = true;
+    this.loaded = false;
+    DumbPipe.close(this.sender);
 };
 
 AudioPlayer.prototype.getMediaTime = function() {
-    return Math.round(this.audio.currentTime * 1000);
+    var p =  new Promise(function(resolve, reject) {
+        this.sender({ type: "getMediaTime" });
+        this.messageHandlers.mediaTime.push(function(data) {
+            resolve(data);
+        });
+    }.bind(this));
+    asyncImpl("I", p);
```

Comment:
This change means that *nGetMediaTime* will sometimes be asynchronous, which isn't clear from reading that function, as it appears to return a value synchronously.

To clarify the behavior of these two functions, I'd make *AudioPlayer.getMediaTime* return its promise, and *nGetMediaTime* call *asyncImpl* if the return value of its call to *player.getMediaTime* is a promise, i.e. something like:

```JavaScript
AudioPlayer.prototype.getMediaTime = function() {
    return new Promise(function(resolve, reject) {
        this.sender({ type: "getMediaTime" });
        this.messageHandlers.mediaTime.push(function(data) {
            resolve(data);
        });
    }.bind(this));
};
…
Native["com/sun/mmedia/DirectPlayer.nGetMediaTime.(I)I"] = function(handle) {
    var player = Media.PlayerCache[handle];
    var mediaTime = player.getMediaTime();
    if (mediaTime instanceof Promise) {
        asyncImpl("I", mediaTime);
    } else {
        return mediaTime;
    }
};
```

(I'd do the same for *AudioPlayer.getDuration*/*nGetDuration*.)